File: 300/388 - Longest Absolute File Path.java
import java.util.HashMap;

public class Solution {
    public int lengthLongestPath(String input) {
        HashMap<Integer, Integer> hm = new HashMap<Integer, Integer>();
        int ret = 0, level, len;
        
        hm.put(0, 0);
        
        for (String name : input.split("\n")) {
            level = name.lastIndexOf("\t") + 1;
            len = name.length() - level;
            
            if (name.contains("."))
                ret = Math.max(ret, hm.get(level) + len);
            else
                hm.put(level + 1, hm.get(level) + len + 1);
        }
        
        return ret;
    }
}

File: csharp/388.longest-absolute-file-path.cs
/*
 * @lc app=leetcode id=388 lang=csharp
 *
 * [388] Longest Absolute File Path
 */

using System.Collections.Generic;
using System.Text.RegularExpressions;

public class Solution {
    public int LengthLongestPath(string input) {
        List<int> lengths = new List<int>();
        int ret = 0, length = -1;
        Regex regex = new Regex(@"[A-Za-z0-9]+\.[A-Za-z0-9]+");
        foreach (string dir in input.Split('\n')) {
            int count = 0;
            foreach (char c in dir) {
                if (c != '\t') {
                    break;
                }
                ++count;
            }
            int len = dir.Length - count + 1,  diff = lengths.Count - count;
            while (diff-- > 0) {
                int j = count + diff;
                length -= lengths[j];
                lengths.RemoveAt(j);;
            }
            lengths.Add(len);
            length += len;
            if (regex.Match(dir).Success) {
                ret = Math.Max(ret, length);
            }
        }
        return ret;
    }
}

